base/rational_to_float.jl

```diff
@@ -0,0 +1,214 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+module RationalToFloat
+
+const Rnd = Base.Rounding
+
+# Performance optimization. Unlike raw `<<` or `>>>`, this is supposed
+# to compile to a single instruction, because the semantics correspond
+# to what hardware usually provides.
+function machine_shift(shift::S, a::T, b) where {S,T<:Base.BitInteger}
+    @inline begin
+        mask = 8*sizeof(T) - 1
+        c = b & mask
+        shift(a, c)
+    end
+end
```

Comment:
Is this really needed? It does seem to simplify the generated code. But some very quick benchmarking don't reveal any differences in the final performance to me. With it I get
```
julia> @benchmark Float64($(12376431 // 8673421))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  11.099 ns … 75.558 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.207 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.303 ns ±  1.022 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▃█                                                          
  ▃███▃▂▃▃▂▂▂▁▁▁▂▁▂▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂ ▂
  11.1 ns         Histogram: frequency by time        13.7 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
end
```
and without
```
julia> @benchmark Float64($(12376431 // 8673421))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):  11.014 ns … 162.400 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.136 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   11.387 ns ±   1.870 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▆█▄▃       ▂▂                                                ▁
  ████▇▃▃▄▁▁▁██▇▆▇▆▅▅▆▆▅▆▆▆▆▅▄▅▅▆▅▆▆▅▆▆▇▆▇▇▇▇▇▆▄▄▅▁▃▁▁▁▄▃▅▃▃▄▄ █
  11 ns         Histogram: log(frequency) by time      16.8 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
Any differences seem to be too small to notice here at least.

Reply:
Disregard the above! I had forgotten to reload the code... It does indeed make a big difference for performence! Without it I get
```

julia> @benchmark Float64($(12376431 // 8673421))
BenchmarkTools.Trial: 10000 samples with 998 evaluations.
 Range (min … max):  15.164 ns … 165.161 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     15.519 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   15.681 ns ±   1.703 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

   ▁ ▂▇█▇▂                   ▁▂▂                               ▂
  ▇███████▁▄▃▁▄▃▃▃▄▃▄▄▃▄▄▅▅▅▆███▇▅▆▅▅▅▅▆▅▆▅▅▅▅▆▄▄▅▄▅▅▅▅▃▅▆█▇▅▆ █
  15.2 ns       Histogram: log(frequency) by time      18.9 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
So more than 30% slower!